### ISSUE TITLE
FR: Function to add heads to an existing hydra

### ISSUE THREAD

**Issue by ennokde** [abo-abo/hydra#185]
tl;dr: The Emacs hydra should be able to grow additional heads just like the mythological one. It would help to keep the separation of the configuration of different packages provided by use-/req-package.

**Current:**
I try to keep things clean in my config by using req-package to require and configure all my packages. If I want to have a hydra containing the functionality of several packages right now I have two options which both introduce otherwise unnecessary dependencies.
The first one is to make one package depend on all others involved and define a big hydra in its config block. This forces me to remember this dependency when I want to add/remove/replace a package. And it puts part of the configuration of one package into the config block of another. So it works against the req-package idea of keeping things separate.
The second option is to use inheritance. This makes it possible to put the heads provided by each package into the appropriate config block. But it also forces me to add dependencies to the req-package declarations to make them load in the right order. Otherwise the hydra I want to inherit from might not be defined yet. So this forces a specific loading sequence that makes deferred loading of packages complicated.

**Desired:**
A third option to add heads to an existing hydra would make the configuration a lot easier. One would only have to worry about defining the hydra before the rest of the configuration. After that all the following config blocks can add their heads at their own leisure. Deferred loading and the sequence of loaded packages do not matter anymore.

**Comment by abo-abo** [comment]
> The second option is to use inheritance.

I'd just go with this. Give an example of why this doesn't work.

> A third option to add heads to an existing hydra would make the configuration a lot easier.

It can work:

    (defhydra hydra-test ()
      ("j" forward-char "forward"))
    
    (define-key hydra-test/keymap "k"
      (eval (hydra--make-defun
             'hydra-test nil nil '("k" backward-char)
             'hydra-test/keymap nil nil)))
             
But I don't think it's worth it. You could wrap the above in something like `(hydra-add-head 'hydra-test '("k" backward-char))` though.

**Comment by ennokde** [comment]
Sorry for the late answer.
Yes, I could use inheritance to do this just as I described. I can express that easily with one added line in the req-package statement:

`(req-package PackageB
:require PackageA
:config
(defhydra hydra-test (:inherit (hydra-PackageA/heads))
  ("j" forward-char "forward"))`

The reason I would like to avoid it is that it would force the packages involved to load in a specific order.

**Comment by ennokde** [comment]
Sorry for the late answer.
It is not impossible but it seems rather unclean. I am still quite new to Emacs after a few months of light usage so correct me if I overlooked something.

Example:
Suppose I have a hydra A defined at init and some packages loaded later by some use- or req-package automation and I want each of those packages to add heads to the hydra. If I used inheritance I would make a new hydra B inheriting from A and another hydra C inheriting from B and so on for the deferred packages so all functions end up in the same hydra.
Now when package C should be loaded B would also have to be loaded before that because I made the C hydra inherit from B to get the heads of A.

I would like to just define a (possibly empty) hydra before any other packages load and then put some `(hydra-add-head ...)` lines into the req-package blocks of other packages of which I want functions in that hydra.

A `(hydra-add-head)` can remove the self-made dependency introduced by the inheritance. That's why I think it might be a good idea to have it. It sounds a bit cleaner to me than the other two approaches.
When I remove the req-package block for a package I also get rid of the heads in the hydra without looking in other places or fixing an inheritance chain.

**Comment by 4lex1v** [comment]
I'm also interested in this, if i got the original request correctly. I'm using use-package and was thinking about adding a custom `:hydra` keyword to the configuration, e.g:

``` lisp
:hydra (:global ("m" "magit")
        :config (:color blue)
        :keys (("s" magit-status          "status")
               ("r" magit-show-refs-popup "branches")))
```

Similar to the existing `:bind` functionality, which would generate `defhydra` and "inject" this hydra into my global hydra.

**Comment by noctuid** [comment]
Could you consider adding something like this directly to hydra? To me, hydra's biggest shortcoming is that all bindings must be in the same location (even if they are from different packages, for example). I think it would be better if it was more similar to `define-prefix-command`. You could define the hydra and then later add the desired keybindings. It would be nice if there was something like a `hydra-add-head` function (not macro). For example:

`(hydra-add-head test-hydra/keymap "k" 'backward-char "back" :keyword val)`
or
`(hydra-add-head 'test-hydra "k" 'backward-char "back" :keyword val)`

It would accept all the local keywords that heads can accept and would use the defaults from the `defhydra` body. Unlike the above example, it would be equivalent to if the keys were in the original hydra. For example, the doc/hint would be regenerated, and it would reuse the original `body-pre`, `body-before-exit`, etc. (these would have to be recorded by defhydra). 

Unfortunately it looks like this would require a fair bit of refactoring.

**Comment by abo-abo** [comment]
> Unfortunately it looks like this would require a fair bit of refactoring.

A small bit, I think. I think `hydra-add-head` can be a simple combination of `define-key` and `hydra--make-defun`. The only thing remaining for `hydra-add-head` is to massage its args into a format `hydra--make-defun` accepts.

**Comment by jkitchin** [comment]
I am trying to add a head to an existing hydra following this thread. This code adds a keybinding to the hydra that opens my email.

```
(define-key scimax-applications/keymap "e"
  (eval (hydra--make-defun
         'scimax-applications nil nil
	 '("e" (if (get-buffer "*mu4e-headers*") (progn (switch-to-buffer
							 "*mu4e-headers*")
							(delete-other-windows))
		 (mu4e)))
	 'scimax-applications/keymap nil nil)))
```

but, it doesn't act like a head that closes the hydra, so it stays open. Is there some trick I need to add at the end to close the hydra without calling any other functions?

**Comment by abo-abo** [comment]
Please test if this API works for you:

```elisp
(defhydra hydra-test-1 ()
  ("j" forward-char "forward")
  ("k" backward-char "backward"))

(defhydra hydra-test-2 ()
  ("j" forward-char "forward"))

(defmacro hydra-add-heads (hydra &rest heads)
  (declare (indent 1))
  `(defhydra ,hydra ()
     ,@(cl-delete-duplicates
        (append
         (symbol-value (intern (concat (symbol-name hydra) "/heads")))
         heads)
        :key #'car :test #'equal)))

(hydra-add-heads hydra-test-2
  ("k" backward-char "backward"))

(hydra-add-heads hydra-test-2
  ("k" backward-char "backward" :exit t))
```

If no issues are found, I'll add this to hydra.el.

**Comment by jkitchin** [comment]
This does add the head, but seems to clobber the hint string.

```
(defmacro hydra-add-heads (hydra &rest heads)
  (declare (indent 1))
  `(defhydra ,hydra ()
     ,(symbol-value (intern (concat (symbol-name hydra) "/hint")))
     ,@(cl-delete-duplicates
        (append
         (symbol-value (intern (concat (symbol-name hydra) "/heads")))
         heads)
        :key #'car :test #'equal)))
```
seems to preserve the hint string I made for it. I have played a little with modifying the hint-string, but it is a little clunky so far.

Here is the full snippet that does what I want for now:

```
(defmacro hydra-add-heads (hydra hint-string &rest heads)
  (declare (indent 1))
  `(defhydra ,hydra ()
     ,(or hint-string (symbol-value (intern (concat (symbol-name hydra) "/hint"))))
     ,@(cl-delete-duplicates
        (append
         (symbol-value (intern (concat (symbol-name hydra) "/heads")))
         heads)
        :key #'car :test #'equal)))

(hydra-add-heads scimax-applications
  nil
  ("e" (progn (if (get-buffer "*mu4e-headers*")
		  (progn (switch-to-buffer
			  "*mu4e-headers*")
			 (delete-other-windows))
		(mu4e)))
   "email" :hint nil :exit t))


;; Add email to the hint string
(setf (second (second scimax-applications/hint))
      (concat (second (second scimax-applications/hint)) "%s: email\n"))


(setf (rest (cdr  (second scimax-applications/hint)))
      (append (rest (cdr  (second scimax-applications/hint))) (list (propertize "e" 'face 'hydra-face-blue))))
```

Thanks for the hints!

**Comment by jkitchin** [comment]
I favor flexibility.  `defhydra+` seems indicative of what it is doing. `defhydra++`?

I think the docstring should be modfiable somehow; it won't be long before someone runs into it. Ditto for body (I assume this is where you could change the color of the head, etc?). 

Is the macro above using some brand new hydra code? I don't have a hydra--prop command.

**Comment by abo-abo** [comment]
Added `defhydra+`. Please test.
